Select element Interaction Regions don't align with other inputs in including their padding

yoelhawa opened the pull request:
#### fea08da84e651a0fc6f1b4862f139fc2f0b0ee17
<pre>
Select element Interaction Regions don&apos;t align with other inputs in including their padding
<a href="https://bugs.webkit.org/show_bug.cgi?id=271499">https://bugs.webkit.org/show_bug.cgi?id=271499</a>
&lt;<a href="https://rdar.apple.com/114213996">rdar://114213996</a>&gt;

Reviewed by Tim Horton.

During the EventRegion paint phase, ignore control clipping to ensure
that Interaction Regions are properly sized.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::pushContentsClip):
Set `isControlClip` to false if we are in the EventRegion paint phase.

* LayoutTests/interaction-region/select-element-expected.txt: Added.
* LayoutTests/interaction-region/select-element.html: Added.
Add a test to cover `&lt;select&gt;` elements&apos; Interaction Regions.
* LayoutTests/interaction-region/text-controls-expected.txt:
* LayoutTests/interaction-region/text-controls.html:
* LayoutTests/interaction-region/text-input-expected.txt:
* LayoutTests/interaction-region/text-input.html:
* LayoutTests/interaction-region/button-in-link.html:
* LayoutTests/interaction-region/button-in-link-expected.txt:
Add test cases with increased borders and overflows to cover this change.
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
Update test expectations to include 1px border on file inputs.
* LayoutTests/platform/ios/editing/editable-region/overflow-scroll-text-field-and-contenteditable-expected.txt:
* LayoutTests/platform/ios/editing/editable-region/search-field-basic-expected.txt:
* LayoutTests/platform/ios/editing/editable-region/text-field-basic-expected.txt:
* LayoutTests/platform/ios/editing/editable-region/text-field-inside-composited-negative-z-index-layer-expected.txt:
Re-baseline iOS Event Region tests.

Canonical link: <a href="https://commits.webkit.org/277387@main">https://commits.webkit.org/277387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23de2cfe8135049427ebcd46bc1cd4d4f8c8acc8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46170 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25301 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48755 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48841 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42211 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48477 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22744 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37726 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46748 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22372 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39808 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18932 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19765 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40906 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4214 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42469 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41255 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50651 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21166 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17657 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44910 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22466 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43819 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10473 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22825 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22160 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on LayoutTests/interaction-region/text-controls.html:
```diff
@@ -16,6 +16,7 @@
 <body>
 <textarea></textarea>
 <input type="text"></input>
+<input type="text" style="border: 10px solid gray"></input>
```
👍

etiennesegonzac commented on LayoutTests/interaction-region/text-input.html:
```diff
@@ -12,7 +12,7 @@
         display: block;
         border-radius: 0;
-        border: 1px gray solid;
+        border: 5px gray solid;
```
good move, this makes things more obvious in the expectation file

etiennesegonzac reviewed (commented):
Looking great!
Adding reviewers to the PR.

mwyrzykowski commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2166,7 +2166,7 @@ bool RenderBox::pushContentsClip(PaintInfo& paintInfo, const LayoutPoint& accumu
         return false;
 
-    bool isControlClip = hasControlClip();
+    bool isControlClip = paintInfo.phase != PaintPhase::EventRegion && hasControlClip();
```
@etiennesegonzac are you ok with this change?

etiennesegonzac replied:
it makes sense for Interaction Regions, but we need to look at Event Regions on iOS a bit more:
https://ews-build.s3-us-west-2.amazonaws.com/iOS-17-Simulator-WK2-Tests-EWS/9a2818d6-21213/results.html
(not worried about the text field ones, but the anchor-lock.html needs some investigation!)

smfr replied:
Do we understand how big things get without the control clip? Are there controls that can have significant overspill that we just don't happen to have tests for? Try removing it for the normal paint code and eyeball the results for all kinds of controls.

yoelhawa replied:
I already added some tests for select element control overspilling into the new `select-element.html` test file and it has correct respective clipping for both the input and interaction region, but I'll add more tests for other kinds of overspill with other controls.

etiennesegonzac commented on LayoutTests/platform/ios/editing/editable-region/overflow-scroll-text-field-and-contenteditable-expected.txt:
```diff
@@ -11,7 +11,7 @@
         (rect (0,0) width=800 height=600)
       (editable region
-        (rect (10,203) width=153 height=20)
+        (rect (9,202) width=155 height=22)
```
these changes look reasonable but maybe other people will want to weigh-in. @smfr maybe?

smfr replied:
I don't recall what editable regions are used for, so can't comment on whether this is significant. Follow the breadcrumbs from https://bugs.webkit.org/show_bug.cgi?id=209888

Commit:
Select element Interaction Regions don't align with other inputs in including their padding
https://bugs.webkit.org/show_bug.cgi?id=271499
<rdar://114213996>

Reviewed by Tim Horton.

During the EventRegion paint phase, ignore control clipping to ensure
that Interaction Regions are properly sized.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::pushContentsClip):
Set `isControlClip` to false if we are in the EventRegion paint phase.

* LayoutTests/interaction-region/select-element-expected.txt: Added.
* LayoutTests/interaction-region/select-element.html: Added.
Add a test to cover `<select>` elements' Interaction Regions.
* LayoutTests/interaction-region/text-controls-expected.txt:
* LayoutTests/interaction-region/text-controls.html:
* LayoutTests/interaction-region/text-input-expected.txt:
* LayoutTests/interaction-region/text-input.html:
* LayoutTests/interaction-region/button-in-link.html:
* LayoutTests/interaction-region/button-in-link-expected.txt:
Add test cases with increased borders and overflows to cover this change.
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
Update test expectations to include 1px border on file inputs.
* LayoutTests/platform/ios/editing/editable-region/overflow-scroll-text-field-and-contenteditable-expected.txt:
* LayoutTests/platform/ios/editing/editable-region/search-field-basic-expected.txt:
* LayoutTests/platform/ios/editing/editable-region/text-field-basic-expected.txt:
* LayoutTests/platform/ios/editing/editable-region/text-field-inside-composited-negative-z-index-layer-expected.txt:
Re-baseline iOS Event Region tests.

Canonical link: https://commits.webkit.org/277387@main
